### PR TITLE
[stable] fix(ha): Create sysctl to set DSCP on Carp

### PR DESCRIPTION
**Pull request by darkfiberiru**
fix(ha): Create sysctl to set DSCP CS7(network control traffic) on
    carp traffic. This will allow switches to use qos to avoid dropping
    packets. By default sysctl is set to false not changing behavior
    Ticket: #28395

### CODE REVIEW

File: sys/netinet/ip_carp.c [hunk 1]
```diff
@@ -819,12 +825,17 @@ carp_send_ad_locked(struct carp_softc *sc)
 #ifdef INET
 	if (sc->sc_naddrs) {
 		struct ip *ip;
-
+		uint8_t ip_tos = IPTOS_DSCP_CS7;
+		
+
```
Comment (amotin) [inline sys/netinet/ip_carp.c]: You've replaced one blank line with two tab-filled lines.

File: sys/netinet/ip_carp.c [hunk 2]
```diff
@@ -819,12 +825,17 @@ carp_send_ad_locked(struct carp_softc *sc)
...
+		
+		
 		m = m_gethdr(M_NOWAIT, MT_DATA);
 		if (m == NULL) {
 			CARPSTATS_INC(carps_onomem);
 			goto resched;
 		}
+                if (!V_carp_dscp) {
+			ip_tos = IPTOS_LOWDELAY;
+		}
```
Comment (amotin) [inline sys/netinet/ip_carp.c]: The curved brackets are not needed here.  And "else" here IMHO could be slightly more readable then assigning the other value above.

File: sys/netinet/ip_carp.c [hunk 3]
```diff
@@ -190,6 +190,10 @@ static int proto_reg[] = {-1, -1};
 static VNET_DEFINE(int, carp_allow) = 1;
 #define	V_carp_allow	VNET(carp_allow)
 
+/* Set DSCP in outgoing CARP packets. */
+static VNET_DEFINE(int, carp_dscp) = 0;
```
Comment (amotin) [inline sys/netinet/ip_carp.c]: Is there any reason to use that default?  I am fine about providing sysctl for people who with to stay back, but why should not default to go forward?

File: sys/netinet/ip_carp.c [hunk 4]
```diff
@@ -886,6 +897,13 @@ carp_send_ad_locked(struct carp_softc *sc)
 		ip6 = mtod(m, struct ip6_hdr *);
 		bzero(ip6, sizeof(*ip6));
 		ip6->ip6_vfc |= IPV6_VERSION;
+                if (V_carp_dscp) {
+			uint8_t LEN_FLOW_LABEL = 20;
```
Comment (amotin) [inline sys/netinet/ip_carp.c]: Why variable instead of macro?

File: sys/netinet/ip_carp.c [hunk 5]
```diff
@@ -886,6 +897,13 @@ carp_send_ad_locked(struct carp_softc *sc)
...
+			/* Traffic class isn't defined in ip6 struct instead  
+			 * it gets offset into flowid field */
+			uint32_t ip_tos = IPTOS_DSCP_CS7 << LEN_FLOW_LABEL;
+                	ip6->ip6_flow |= ip_tos;
```
Comment (amotin) [inline sys/netinet/ip_carp.c]: Why another variable is needed here at all?  It does not make things more compact or clear.

And what's about the byte order?  Shouldn't here be some htonl()?

**Review by amotin — COMMENTED** [review]
It would be good to push that to FreeBSD's phabricator for review of network people.

File: sys/netinet/ip_carp.c (hunk 4 above)
Comment (darkfiberiru) [reply sys/netinet/ip_carp.c]: What kind of macro would you like here? We don't have that defined anywhere that I could find?

Comment (amotin) [reply sys/netinet/ip_carp.c]: I don't know whether there is existing define for that, but if not, why not define it here, why to use the variable?

**Comment by darkfiberiru** [comment]
Still need to figure out/verify network ordering endian fun. Cleaned up a few other stylistic things and made this the new default. Need to make sure this enters change log clearly as we don't have a dot updating file like freebsd does.

**Comment by darkfiberiru** [comment]
Added to reviews.freebsd.org
https://reviews.freebsd.org/D14536

**Review by amotin — APPROVED** [review]

**Comment by darkfiberiru** [comment]
Messed up tree redoing pull request

**Pull request closed without merging**